File: handicraft_export/handicraft/report/artisan_yield_wastage/artisan_yield_wastage.js
// Copyright (c) 2026, Your Company and contributors
// For license information, please see license.txt

frappe.query_reports["Artisan Yield & Wastage"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
		},
		{
			"fieldname": "artisan",
			"label": __("Artisan"),
			"fieldtype": "Link",
			"options": "Artisan",
		},
		{
			"fieldname": "artisan_cluster",
			"label": __("Artisan Cluster"),
			"fieldtype": "Link",
			"options": "Artisan Cluster",
		},
	],
	"formatter": function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname === "yield_pct" && data && data.yield_pct !== undefined) {
			if (data.yield_pct < 80) {
				value = `<span style="color:red;font-weight:bold;">${value}</span>`;
			} else if (data.yield_pct < 95) {
				value = `<span style="color:orange;font-weight:bold;">${value}</span>`;
			} else {
				value = `<span style="color:green;font-weight:bold;">${value}</span>`;
			}
		}
		return value;
	},
};

File: handicraft_export/handicraft/report/artisan_yield_wastage/artisan_yield_wastage.py
# Copyright (c) 2026, Your Company and contributors
# For license information, please see license.txt

from __future__ import unicode_literals

import frappe
from frappe import _


def execute(filters=None):
	columns = get_columns()
	data = get_data(filters)
	return columns, data


def get_columns():
	return [
		{
			"fieldname": "artisan_name",
			"label": _("Artisan"),
			"fieldtype": "Data",
			"width": 150,
		},
		{
			"fieldname": "artisan_cluster",
			"label": _("Cluster"),
			"fieldtype": "Data",
			"width": 120,
		},
		{
			"fieldname": "job_card",
			"label": _("Job Card"),
			"fieldtype": "Link",
			"options": "Artisan Job Card",
			"width": 140,
		},
		{
			"fieldname": "export_order",
			"label": _("Export Order"),
			"fieldtype": "Link",
			"options": "Export Order",
			"width": 140,
		},
		{
			"fieldname": "qty_ordered",
			"label": _("Qty Ordered"),
			"fieldtype": "Int",
			"width": 100,
		},
		{
			"fieldname": "qty_received",
			"label": _("Qty Received"),
			"fieldtype": "Int",
			"width": 100,
		},
		{
			"fieldname": "qty_rejected",
			"label": _("Qty Rejected"),
			"fieldtype": "Int",
			"width": 100,
		},
		{
			"fieldname": "qty_wastage",
			"label": _("Wastage"),
			"fieldtype": "Int",
			"width": 80,
		},
		{
			"fieldname": "yield_pct",
			"label": _("Yield %"),
			"fieldtype": "Percent",
			"width": 90,
		},
	]


def get_data(filters):
	conditions = []
	values = {}

	if filters.get("from_date"):
		conditions.append("jc.modified >= %(from_date)s")
		values["from_date"] = filters["from_date"]

	if filters.get("to_date"):
		conditions.append("jc.modified <= %(to_date)s")
		values["to_date"] = filters["to_date"]

	if filters.get("artisan"):
		conditions.append("jc.artisan = %(artisan)s")
		values["artisan"] = filters["artisan"]

	if filters.get("artisan_cluster"):
		conditions.append("a.artisan_cluster = %(artisan_cluster)s")
		values["artisan_cluster"] = filters["artisan_cluster"]

	where_clause = " AND ".join(conditions) if conditions else "1=1"

	query = f"""
		SELECT
			a.artisan_name,
			ac.cluster_name AS artisan_cluster,
			jc.name AS job_card,
			jc.export_order,
			jc.qty_ordered,
			jc.qty_received,
			jc.qty_rejected,
			(jc.qty_ordered - jc.qty_received) AS qty_wastage,
			CASE
				WHEN jc.qty_ordered > 0
				THEN ROUND((jc.qty_received / jc.qty_ordered) * 100, 2)
				ELSE 0
			END AS yield_pct
		FROM `tabArtisan Job Card` jc
		INNER JOIN `tabArtisan` a ON a.name = jc.artisan
		LEFT JOIN `tabArtisan Cluster` ac ON ac.name = a.artisan_cluster
		WHERE jc.docstatus = 1 AND {where_clause}
		ORDER BY yield_pct ASC
	"""

	return frappe.db.sql(query, values, as_dict=True)
